tests/contracts/test_writer_publication_preview_contracts.py

```
@@ -0,0 +1,144 @@
+        self.validate("manuscript-package.schema.json",f["manuscript_package"]); self.validate("publication-preview.schema.json",f["preview_artifact_manifest"]); self.validate("publication-preview.schema.json",f["preview_iteration"])
+        for d in f["profile_defect_candidates"]: self.validate("publication-preview.schema.json",d)
+
+    def test_error_catalog_exact(self): self.assertEqual(set(self.sem["errors"]), ERROR_IDS)
+
+    def test_research_package_binding_and_firewall(self):
+        p=self.f["research_package"]; self.assertIsNone(research_package_error(p))
+        b=deepcopy(p); b["release_eligible"]=True; refresh(b,"package_digest"); self.assertEqual(research_package_error(b),"WP-EPISTEMIC-FIREWALL-001")
+        b=deepcopy(p); b["provenance"]["input_digests"].remove(p["source_research_snapshot"]["content_digest"]); refresh(b,"package_digest"); self.assertEqual(research_package_error(b),"WP-RP-SNAPSHOT-BINDING-001")
+
+    def test_preview_source_namespace_is_bidirectional(self):
+        p=deepcopy(self.f["research_package"]); source=p["content"]["source_refs"][1]
```

Ruff (0.16.2)

[error] Multiple statements on one line (semicolon)

(E702)